tests/unit/s2n_self_talk_client_hello_cb_test.c

```diff
@@ -165,86 +180,52 @@ int client_hello_fail_handshake(struct s2n_connection *conn, void *ctx)
-    EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(swap_config, chain_and_key));
-
-    /* Add application protocols to swapped config */
-    const char *protocols[] = { "h2" };
-    EXPECT_SUCCESS(s2n_config_set_protocol_preferences(swap_config, protocols, 1));
-
-    /* Prepare context */
-    client_hello_ctx.invoked = 0;
-    client_hello_ctx.swap_config = 1;
-    client_hello_ctx.config = swap_config;
-
-    /* Set up the callback to swap config on client hello */
-    EXPECT_SUCCESS(s2n_config_set_client_hello_cb(config, client_hello_swap_config, &client_hello_ctx));
-
-    /* Create a pipe */
-    struct s2n_test_io_pair io_pair;
-    EXPECT_SUCCESS(s2n_io_pair_init(&io_pair));
-
-    /* Create a child process */
-    pid = fork();
-    if (pid == 0) {
-        /* This is the client process, close the server end of the pipe */
-        EXPECT_SUCCESS(s2n_io_pair_close_one_end(&io_pair, S2N_SERVER));
-
-        mock_client(&io_pair, 0, 1);
+    EXPECT_NOT_NULL(conn);
+    /* negotiate handshake, we should pause after the nonblocking callback is invoked */
+    EXPECT_FAILURE_WITH_ERRNO(s2n_negotiate(conn, &blocked), S2N_ERR_ASYNC_BLOCKED);
+    EXPECT_EQUAL(blocked, S2N_BLOCKED_ON_APPLICATION_INPUT);
+    
+    /* verify client hello cb has been invoked */
+    EXPECT_EQUAL(ch_ctx->invoked, 1);
+
+    /* while handshake is paused, swap the config if asked */
+    if (ch_ctx->swap_config_nonblocking_mode) {
+        EXPECT_SUCCESS(s2n_connection_set_config(conn, ch_ctx->config));
     }
+    /* unless explicitly unblocked we should stay paused */
+    EXPECT_FAILURE_WITH_ERRNO(s2n_negotiate(conn, &blocked), S2N_ERR_ASYNC_BLOCKED);
+    EXPECT_EQUAL(blocked, S2N_BLOCKED_ON_APPLICATION_INPUT);    
+
+    /* mark the client hello cb complete */
+    EXPECT_SUCCESS(s2n_client_hello_cb_done(conn));
+    if (server_name_used) {
+        EXPECT_SUCCESS(s2n_connection_server_name_extension_used(conn));
+    }
+    return s2n_negotiate(conn, &blocked);
+}
 
-    /* This is the server process, close the client end of the pipe */
-    EXPECT_SUCCESS(s2n_io_pair_close_one_end(&io_pair, S2N_CLIENT));
-
-    EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_SERVER));
-    EXPECT_SUCCESS(s2n_connection_set_config(conn, config));
-
-    /* Set up the connection to read from the fd */
-    EXPECT_SUCCESS(s2n_connection_set_io_pair(conn, &io_pair));
-
-    /* Negotiate the handshake. */
-    EXPECT_SUCCESS(s2n_negotiate(conn, &blocked));
-
-    /* Server name and error are as expected with null connection */
-    EXPECT_NULL(s2n_get_server_name(NULL));
-    EXPECT_EQUAL(s2n_errno, S2N_ERR_NULL);
-
-    /* Ensure that callback was invoked */
-    EXPECT_EQUAL(client_hello_ctx.invoked, 1);
+int s2n_negotiate_blocking_ch_cb(struct s2n_connection *conn, struct client_hello_context *ch_ctx)
+{
+    s2n_blocked_status blocked;
+    EXPECT_NOT_NULL(conn);
+    
+    return s2n_negotiate(conn, &blocked);
```

Comment:
Just wanted to verify this line, should the rest be removed? Or the `return` be removed?

Reply:
in this case we are just interested in verifying blocking negotiate was successful.

callers use this as: `EXPECT_SUCCESS(s2n_negotiate_blocking_ch_cb(..))`

Reply:
Ryan's point is that you appear to have more code after this return, including another return?

Reply:
fixed the unreachable code.